tests/gdx-tests/src/com/badlogic/gdx/tests/g3d/shadows/system/BaseShadowSystem.java

```diff
@@ -0,0 +1,468 @@
+import com.badlogic.gdx.tests.g3d.shadows.utils.LightFilter;
+import com.badlogic.gdx.tests.g3d.shadows.utils.NearFarAnalyzer;
+import com.badlogic.gdx.tests.g3d.shadows.utils.ShadowMapAllocator;
+import com.badlogic.gdx.tests.g3d.shadows.utils.ShadowMapAllocator.ShadowMapRegion;
+import com.badlogic.gdx.utils.GdxRuntimeException;
+import com.badlogic.gdx.utils.ObjectMap;
+import com.badlogic.gdx.utils.ObjectMap.Entries;
+
+/** BaseShadowSystem allows to easily create custom shadow system. It assumes that the first pass renders all depth maps in packed
+ * texture.
+ * @author realitix */
+public abstract class BaseShadowSystem implements ShadowSystem {
+	/** This class handles camera and texture region.
+	 * @author realitix */
+	public static class LightProperties {
+		public Camera camera;
+		public TextureRegion region = new TextureRegion();
+
+		public LightProperties (Camera camera) {
+			this.camera = camera;
+		}
+	}
+
+	/** This class handles LightProperties for each side of PointLight.
+	 * @author realitix */
+	public static class PointLightProperties {
+		public ObjectMap<CubemapSide, LightProperties> properties = new ObjectMap<CubemapSide, LightProperties>(6);
+	}
+
+	/** First pass number */
+	public static final int PASS_1 = 0;
+
+	/** Main camera */
+	protected Camera camera;
+	/** Renderable providers */
+	protected Iterable<RenderableProvider> renderableProviders;
+	/** Cameras linked with spot lights */
+	protected ObjectMap<SpotLight, LightProperties> spotCameras = new ObjectMap<SpotLight, LightProperties>();
+	/** Cameras linked with directional lights */
+	protected ObjectMap<DirectionalLight, LightProperties> dirCameras = new ObjectMap<DirectionalLight, LightProperties>();
+	/** Cameras linked with point lights */
+	protected ObjectMap<PointLight, PointLightProperties> pointCameras = new ObjectMap<PointLight, PointLightProperties>();
+	/** Analyzer of near and far for spot and point lights */
+	protected NearFarAnalyzer nearFarAnalyzer;
+	/** Allocator which choose where to render shadow map in texture */
+	protected ShadowMapAllocator allocator;
+	/** Analyzer which compute how to create the camera for directional light */
+	protected DirectionalAnalyzer directionalAnalyzer;
+	/** Filter that choose if light must be rendered */
+	protected LightFilter lightFilter;
+	/** Framebuffer used to render all the depth maps */
+	protected FrameBuffer[] frameBuffers;
+	/** Current pass in the depth process */
+	protected int currentPass = -1;
+	/** Iterators for cameras */
+	protected Entries<SpotLight, LightProperties> spotCameraIterator;
+	protected Entries<DirectionalLight, LightProperties> dirCameraIterator;
+	protected Entries<PointLight, PointLightProperties> pointCameraIterator;
+	/** Current side in the point light cubemap */
+	protected int currentPointSide;
+	protected PointLightProperties currentPointProperties;
+	/** Tmp variables */
+	protected Vector3 tmpV3 = new Vector3();
```

Comment:
Why is this here? It doesnt seem to be used. Also why is this protected? You typically want to keep the scope of tmp instances small to avoid problems because you dont know where they are used.